src/lib/optimizer/join_ordering/join_graph_converter.hpp

```diff
@@ -0,0 +1,79 @@
+#pragma once
+
+#include <map>
+#include <memory>
+#include <optional>
+#include <set>
+#include <vector>
+
+#include "join_graph.hpp"
+#include "logical_query_plan/union_node.hpp"
+
+namespace opossum {
+
+class AbstractLQPNode;
+
+/**
+ * Turns an LQP into a JoinGraph.
+ *
+ * Consider this simple LQP:
+ * [0] [Projection] x1, y1
+ *  \_[1] [Predicate] x2 <= y1
+ *     \_[2] [Cross Join]
+ *        \_[3] [MockTable] x
+ *        \_[4] [MockTable] y
+ *
+ * The JoinGraph created from it would contain two vertices (x and y), one edge (between x and y, with the predicate
+ * x2 <= y1) and one parent_relation (Projection, left child side)
+ *
+ * The name "Converter" stems from the fact that nodes representing predicates and joins are physically removed from the
+ * LQP. To turn a JoinGraph back into a LQP, use a JoinOrdering algorithm such as DpCcp.
+ */
+class JoinGraphConverter final {
+ public:
+  /**
+   * From the subtree of root, build a JoinGraph.
+   * The LQP is not modified during this process.
+   *
+   * Need an instance of the shared_ptr to keep the ref count > 0
+   */
+  JoinGraph operator()(const std::shared_ptr<AbstractLQPNode>& lqp);
+
+ private:
+  void _traverse(const std::shared_ptr<AbstractLQPNode>& node);
+
+  /**
+   * A subgraph in the LQP consisting of UnionNodes and PredicateNodes can be translated into a single complex predicate
+   *
+   *         ______________UnionNode________
+   *        /                               \
+   *    PredicateNode(a > 5)            PredicateNode(a < 3)
+   *       |                                 |
+   *       |                            PredicateNode(b > 12)
+   *       \____________AggregateNode_______/
+   *                           |
+   *                         [...]
+   *
+   *  Represents the JoinPlanPredicate "a > 5 OR (a < 3 AND b > 12)". The AggregateNode is the Predicate's "base_node"
+   *  i.e. the node on which's output the Predicate operates on.
```

Comment:
Introducing "base" nodes as an orthogonal concept to "parent/child" leads to ambiguity. Something like "common child"?

(It is becoming more apparent that we didn't do ourselves a favor with the parent/child names)

Reply:
I can call this "common child" but I don't see me introducing a concept here that should affect anyone but the internals of the JoinGraphConverter